SSA/Projects/Scf/Scf.lean

```diff
@@ -401,25 +508,46 @@ def lhs : Com Op [/- v0 -/ t] t :=
   /- start1 = -/ Com.lete (cst start1) <|
   /- c1 = -/ Com.lete (cst 1) <|
   -- start step niter v
-  Com.lete (for_ (t := t) ⟨1, by simp[Ctxt.snoc]⟩ ⟨0, by simp[Ctxt.snoc]⟩ ⟨2, by simp[Ctxt.snoc]⟩ ⟨3, by simp[Ctxt.snoc]⟩ rgn) <|
+  Com.lete (for_ (t := t) ⟨1, by simp [Ctxt.snoc]⟩ ⟨0, by simp [Ctxt.snoc]⟩ ⟨2, by simp [Ctxt.snoc]⟩ ⟨3, by simp [Ctxt.snoc]⟩ rgn) <|
   /- niters2 = -/ Com.lete (cst_nat niters2) <|
   /- start2 = -/ Com.lete (cst <| niters1 + start1) <|
   /- c1 = -/ Com.lete (cst 1) <|
-  Com.lete (for_ (t := t) ⟨1, by simp[Ctxt.snoc]⟩ ⟨0, by simp[Ctxt.snoc]⟩ ⟨2, by simp[Ctxt.snoc]⟩ ⟨3, by simp[Ctxt.snoc]⟩ rgn) <|
-  Com.ret ⟨0, by simp[Ctxt.snoc]⟩
+  Com.lete (for_ (t := t) ⟨1, by simp [Ctxt.snoc]⟩ ⟨0, by simp [Ctxt.snoc]⟩ ⟨2, by simp [Ctxt.snoc]⟩ ⟨3, by simp [Ctxt.snoc]⟩ rgn) <|
+  Com.ret ⟨0, by simp [Ctxt.snoc]⟩
 
 def rhs : Com Op [/- v0 -/ t] t :=
   /- niters1 + niters2 = -/ Com.lete (cst_nat <| niters1 + niters2) <|
   /- start1 = -/ Com.lete (cst start1) <|
   /- c1 = -/ Com.lete (cst 1) <|
   -- start step niter v
-  Com.lete (for_ (t := t) ⟨1, by simp[Ctxt.snoc]⟩ ⟨0, by simp[Ctxt.snoc]⟩ ⟨2, by simp[Ctxt.snoc]⟩ ⟨3, by simp[Ctxt.snoc]⟩ rgn) <|
-  Com.ret ⟨0, by simp[Ctxt.snoc]⟩
-#check rhs
+  Com.lete (for_ (t := t) ⟨1, by simp [Ctxt.snoc]⟩ ⟨0, by simp [Ctxt.snoc]⟩ ⟨2, by simp [Ctxt.snoc]⟩ ⟨3, by simp [Ctxt.snoc]⟩ rgn) <|
+  Com.ret ⟨0, by simp [Ctxt.snoc]⟩
+
 
 theorem correct :
-  Com.denote (lhs rgn niters1 ninters2 start1) Γv = Com.denote (rhs rgn niters1 ninters2 start1) Γv := by
-    sorry -- TODO: finish proof
+  Com.denote (lhs rgn niters1 niters2 start1) Γv = Com.denote (rhs rgn niters1 niters2 start1) Γv := by
+    simp [lhs, rhs, for_, axpy, cst]
+    try simp (config := {decide := false}) only [
+    Com.denote, Expr.denote, HVector.denote, Var.zero_eq_last, Var.succ_eq_toSnoc,
+    Ctxt.empty, Ctxt.empty_eq, Ctxt.snoc, Ctxt.Valuation.nil, Ctxt.Valuation.snoc_last,
+    Ctxt.ofList, Ctxt.Valuation.snoc_toSnoc,
+    HVector.map, HVector.toPair, HVector.toTuple, OpDenote.denote, Expr.op_mk, Expr.args_mk,
+    Function.comp, Ctxt.Valuation.ofPair, Ctxt.Valuation.ofHVector, Function.uncurry,
+    add, cst_nat,
+    Ctxt.Valuation.snoc_eval_one, Ctxt.Valuation.snoc_snoc_eval_two, Ctxt.Valuation.snoc_snoc_snoc_eval_three]
+    have swap_niters := add_comm (a := niters1) (b := niters2)
+    have H : (loop_counter_decorator 1 fun i v =>
+          Com.denote rgn (Ctxt.Valuation.snoc (Ctxt.Valuation.snoc default v) i))^[niters1 + niters2]
+      (start1, Γv (Var.last [] t)) = (loop_counter_decorator 1 fun i v =>
+          Com.denote rgn (Ctxt.Valuation.snoc (Ctxt.Valuation.snoc default v) i))^[niters2 + niters1]
+      (start1, Γv (Var.last [] t)) := by
+        congr
+    simp [H]
+    simp [Function.iterate_add_apply]
+    congr
+    rw [loop_counter_decorator_iterate_fst_val]
+    linarith
+    done
```

Comment:
This proof feels like it could use some golfing/simplifying

Reply:
@alexkeizer Proof has been correctly golfed now!

Reply:
Golfed!

Reply:
Ah, sorry, I missed your original message. In general, I think we ought to have a closer look at `simp_peephole`. The current implementation was nice as a proof of concept, but it's way too brittle.